tests/unit/test_topology.py

```diff
@@ -0,0 +1,112 @@
+from copy import deepcopy
+
+import pytest
+
+from expertsystem.topology import (
+    Edge,
+    InteractionNode,
+    SimpleStateTransitionTopologyBuilder,
+    Topology,
+)
+
+
+class TestEdge:
+    @staticmethod
+    def test_get_connected_nodes():
+        edge = Edge(1, 2)
+        assert edge.get_connected_nodes() == {1, 2}
+        edge = Edge(originating_node_id=3)
+        assert edge.get_connected_nodes() == {3}
+        edge = Edge(ending_node_id=4)
+        assert edge.get_connected_nodes() == {4}
+
+
+class TestTopology:
+    @staticmethod
+    def test_constructor_exceptions():
+        with pytest.raises(ValueError):
+            assert Topology(edges={0: Edge(None, 1)})
+        with pytest.raises(ValueError):
+            assert Topology(nodes={0}, edges={0: Edge(1, None)})
+        with pytest.raises(ValueError):
+            assert Topology(edges={0: Edge(None, None)})
+        topology = Topology(nodes={0}, edges={0: Edge(0, None)})
+        assert len(topology.nodes) == 1
+
+    @staticmethod
+    def test_is_valid(dummy_topology):
+        topology = deepcopy(dummy_topology)
+        assert topology.is_valid()
+        topology.add_node(2)
+        assert not topology.is_valid()
+        topology.add_edges([5])
+        assert not topology.is_valid()
+        topology.attach_edges_to_node_ingoing([5], 2)
+        assert not topology.is_valid()
+
+    @staticmethod
+    def test_check_isolated_nodes(dummy_topology):
```

Comment:
It would propose to remove `is_valid` and `verify` will be the only public interface function (regarding verification). Then various toplogies can be created and verified. An depending on the specifc Errors raised a more granular way of testing is possible.
That means this test can be "merged" with the one above

Reply:
81bc9f5f

Reply:
But it might be better to make the `dummy_topology` a `three_body_topology` and tests a few more of the `Topology` methods

Reply:
Yeah that's def a better name for it

Reply:
38464a68 and bf1b1e66